enhancements/network/baremetal-ipi-network-configuration.md

```diff
@@ -0,0 +1,187 @@
+---
+title: baremetal-ipi-network-configuration
+authors:
+  - "@cybertron"
+  - "@hardys"
+reviewers:
+  - "@kirankt"
+  - "@dtantsur"
+  - "@zaneb"
+approvers:
+  - "TBD"
+creation-date: 2021-05-21
+last-updated: 2021-06-28
+status: provisional|implementable
+
+see-also:
+  - "/enhancements/host-network-configuration.md"
+  - "/enhancements/machine-config/mco-network-configuration.md"
+  - "/enhancements/machine-config/rhcos/static-networking-enhancements.md"
+
+# Baremetal IPI Network Configuration
+
+Describe user-facing API for day-1 network customizations in the IPI workflow,
+with particular focus on baremetal where such configuration is a common
+requirement.
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Currently in the IPI flow, there is no way to provide day-1 network configuration
+which is a common requirement, particularly for baremetal users.  We can build
+on the [UPI static networking enhancements](rhcos/static-networking-enhancements.md)
+to enable such configuration in the IPI flow.
+
+## Motivation
+
+Since the introduction of baremetal IPI, a very common user request is how
+to configure day-1 networking, and in particular the following cases which are not currently possible:
+
+* Deploy with OpenShift Machine network on a tagged (non-default) VLAN
+* Deploy with OpenShift Machine network using static IPs (no DHCP)
+
+In both cases, this configuration cannot be achieved via DHCP so some
+means of providing the configuration to the OS is required.
+
+In the UPI flow this is achieved by consuming user-provided NetworkManager
+keyfiles, as an input to `coreos-install --copy-network`, but there is
+no corresponding user interface at the openshift-install level.
+
+### Goals
+
+* Define API for day-1 network customizations
+* Enable common on-premise network configurations (bond+vlan, static ips) via IPI
+
+### Non-Goals
+
+* Platforms other than `baremetal`, although the aim is a solution which could be applied to other platforms in future if needed.
+* Provide a consistent (ideally common) user API for deployment and post-deployment configuration. Getting agreement on a common API for day-1 and day-2 has proven contentious and will be addressed later.
+* Enabling kubernetes-nmstate by default for day-2 networking is discussed via
+[another proposal](https://github.com/openshift/enhancements/pull/747)
+* Handling of advanced networking options such as bridges and bonds. The implementation may support them, but they are not the priority for this work.
```

Comment:
I'm not sure this is correct - the #1 request for baremetal is bond+vlan, since it's the cheapest way to achieve both a dedicated provisioning network and a highly available MachineNetwork.

Maybe this should say something like "Advanced network configuration which is not achievable via simple network manager keyfiles?"

Reply:
Fully agree with Steve on the need for bond and VLAN as the basic most required config (obviously, along with static IPs).

Reply:
@cybertron oh ok, so the workflow you have in mind could be basic host network config on day 1 and then have it fully configured right after finishing the install with knmstate? If this made sense, we could pass knmstate configurations with the installer to be applied once the knmstate operator is deployed (and automate its deployment as part of the install workflow). On the flip side, users would have to deal with a combination of NetworkManager keyfiles and nmstate, unless some conversion of sorts could be implemented.

Reply:
@racedo ideally we'd pass in the same data consumed by kubernetes-nmstate to the installer (indeed that was our [original plan](https://github.com/hardys/enhancements/blob/day1_networking/enhancements/installer/network_configuration.md))

However discussion on #747 has stalled, and I think based on conversations with @dhellmann and @derekwaynecarr we're not getting knmstate (and associated NNCP CRD API) into the release payload.

So, we need an interim solution which does mean users have to deal with a combination of keyfiles and nmstate.

There is a conversion possible using the nmstate CLI (which @cybertron mentions in this PR), so I think we'll just have to document that until a more fully-integrated solution is possible (either adding knmstate to the release payload, or more likely adding some way to automatically enable OLM operators at install-time)